core/backend/src/internal/workspace/WorkspaceImpl.ts

```diff
@@ -0,0 +1,794 @@
+    this.performWriteSql(rscName, "DELETE FROM strings WHERE id=?");
+  }
+  public addBlob(rscName: WorkspaceResourceName, val: Uint8Array): void {
+    EditableDbImpl.validateResourceName(rscName);
+    this.validateResourceSize(val);
+    this.performWriteSql(rscName, "INSERT INTO blobs(id,value) VALUES(?,?)", (stmt) => stmt.bindBlob(2, val));
+  }
+  public updateBlob(rscName: WorkspaceResourceName, val: Uint8Array): void {
+    this.validateResourceSize(val);
+    this.performWriteSql(rscName, "INSERT INTO blobs(id,value) VALUES(?,?) ON CONFLICT(id) DO UPDATE SET value=excluded.value WHERE value!=excluded.value", (stmt) => stmt.bindBlob(2, val));
+  }
+  public getBlobWriter(rscName: WorkspaceResourceName): SQLiteDb.BlobIO {
+    return this.sqliteDb.withSqliteStatement("SELECT rowid from blobs WHERE id=?", (stmt) => {
+      stmt.bindString(1, rscName);
+      const blobWriter = SQLiteDb.createBlobIO();
+      blobWriter.open(this.sqliteDb.nativeDb, { tableName: "blobs", columnName: "value", row: stmt.getValueInteger(0), writeable: true });
+      return blobWriter;
+    });
+  }
+  public removeBlob(rscName: WorkspaceResourceName): void {
+    this.performWriteSql(rscName, "DELETE FROM blobs WHERE id=?");
+  }
+  public addFile(rscName: WorkspaceResourceName, localFileName: LocalFileName, fileExt?: string): void {
+    EditableDbImpl.validateResourceName(rscName);
+    fileExt = fileExt ?? extname(localFileName);
+    if (fileExt?.[0] === ".")
+      fileExt = fileExt.slice(1);
+    this.sqliteDb.nativeDb.embedFile({ name: rscName, localFileName, date: this.getFileModifiedTime(localFileName), fileExt });
+  }
+  public updateFile(rscName: WorkspaceResourceName, localFileName: LocalFileName): void {
+    this.queryFileResource(rscName); // throws if not present
+    this.sqliteDb.nativeDb.replaceEmbeddedFile({ name: rscName, localFileName, date: this.getFileModifiedTime(localFileName) });
+  }
+  public removeFile(rscName: WorkspaceResourceName): void {
+    const file = this.queryFileResource(rscName);
+    if (undefined === file)
+      throw new Error(`file resource "${rscName}" does not exist`);
+    if (file && fs.existsSync(file.localFileName))
+      fs.unlinkSync(file.localFileName);
+    this.sqliteDb.nativeDb.removeEmbeddedFile(rscName);
+  }
+}
+
+export function constructWorkspaceDb(props: WorkspaceDbProps, container: WorkspaceContainer): WorkspaceDb {
+  return new WorkspaceDbImpl(props, container);
+}
+
+export function constructWorkspace(settings: Settings, opts?: WorkspaceOpts): OwnedWorkspace {
+  return new WorkspaceImpl(settings, opts);
+}
+
+export function constructWorkspaceEditor(): WorkspaceEditor {
+  return new EditorImpl();
+}
+
+export function noLeadingOrTrailingSpaces(name: string, msg: string) {
+  if (name.trim() !== name)
+    throw new Error(`${msg} [${name}] may not have leading or trailing spaces`);
+}
+
+export function validateWorkspaceDbName(dbName: WorkspaceDbName) {
```

Comment:
Can I summarize this as "must be a valid Windows and Unix filename no longer than 255 characters"? It's not currently documented anywhere.

Reply:
I added documention on `WorkspaceDbName`.

Reply:
It came from [here](https://www.npmjs.com/package/filename-reserved-regex)